internal/runbits/runtime/rationalize.go

```diff
@@ -72,13 +67,34 @@ func rationalizeError(auth *authentication.Auth, proj *project.Project, rerr *er
 			break // it only takes one download failure to report the runtime failure as due to download error
 		}
 
+	// We communicate buildplanner errors verbatim as the intend is that these are curated by the buildplanner
+	case errors.As(*rerr, &buildPlannerErr):
+		*rerr = errs.WrapUserFacing(*rerr,
+			buildPlannerErr.LocalizedError(),
+			errs.SetIf(buildPlannerErr.InputError(), errs.SetInput()))
+
+	// User has modified the buildscript and needs to run `state commit`
+	case errors.Is(*rerr, runtime.NeedsCommitError):
+		*rerr = errs.WrapUserFacing(*rerr, locale.T("notice_commit_build_script"), errs.SetInput())
+
+	// Buildscript is missing and needs to be recreated
+	case errors.Is(*rerr, runtime.NeedsBuildscriptResetError):
+		*rerr = errs.WrapUserFacing(*rerr, locale.T("notice_needs_buildscript_reset"), errs.SetInput())
+
+	// Artifact build errors
+	case errors.As(*rerr, &artifactErr):
+		errMsg := locale.Tr("err_build_artifact_failed_msg", artifactErr.Artifact.DisplayName)
+		*rerr = errs.WrapUserFacing(*rerr, locale.Tr("err_build_artifact_failed", errMsg,
+			strings.Join(artifactErr.Artifact.Errors, "\n"), artifactErr.Artifact.LogURL))
+
 	// If updating failed due to unidentified errors, and the user is not authenticated, add a tip suggesting that they authenticate as
 	// this may be a private project.
 	// Note since we cannot assert the actual error type we do not wrap this as user-facing, as we do not know what we're
 	// dealing with so the localized underlying errors are more appropriate.
-	case isUpdateErr && !auth.Authenticated():
+	case !auth.Authenticated(): // MUST BE LAST
```

Comment:
Because this is not asserting an error, it's decorating an unknown error. And all the other assertions are for asserting an error.

I've updated the code with a comment and some protection against accidental refactoring.

Note the original implementation was bugged:

https://github.com/ActiveState/cli/pull/3199/files/3c9e66d2268323d48391a6b19658bd20b4a7e327#diff-ad4a4ce35107eceb3ee56dfede324f0f7eb85af84005dbd6dc158f5681e58c92L28

It was ignoring errors if those non-error assertions were evaluated. We really should void using rationalizers to do anything that's not error assertion related.